src/ray/core_worker/core_worker.cc

```diff
@@ -1250,184 +1249,24 @@ void CoreWorker::RemoveActorHandleReference(const ActorID &actor_id) {
-        direct_actor_submitter_->DisconnectActor(actor_id, true);
-        // We cannot erase the actor handle here because clients can still
-        // submit tasks to dead actors. This also means we defer unsubscription,
-        // otherwise we crash when bulk unsubscribing all actor handles.
-      } else {
-        direct_actor_submitter_->ConnectActor(actor_id, actor_data.address());
-      }
-
-      const auto &actor_state = gcs::ActorTableData::ActorState_Name(actor_data.state());
-      RAY_LOG(INFO) << "received notification on actor, state: " << actor_state
-                    << ", actor_id: " << actor_id
-                    << ", ip address: " << actor_data.address().ip_address()
-                    << ", port: " << actor_data.address().port() << ", worker_id: "
-                    << WorkerID::FromBinary(actor_data.address().worker_id())
-                    << ", raylet_id: "
-                    << ClientID::FromBinary(actor_data.address().raylet_id());
-    };
-
-    RAY_CHECK_OK(gcs_client_->Actors().AsyncSubscribe(
-        actor_id, actor_notification_callback, nullptr));
-
-    RAY_CHECK(reference_counter_->SetDeleteCallback(
-        actor_creation_return_id,
-        [this, actor_id, is_owner_handle](const ObjectID &object_id) {
-          if (is_owner_handle) {
-            // If we own the actor and the actor handle is no longer in scope,
-            // terminate the actor. We do not do this if the GCS service is
-            // enabled since then the GCS will terminate the actor for us.
-            if (!(RayConfig::instance().gcs_service_enabled() &&
-                  RayConfig::instance().gcs_actor_service_enabled())) {
-              RAY_LOG(INFO) << "Owner's handle and creation ID " << object_id
-                            << " has gone out of scope, sending message to actor "
-                            << actor_id << " to do a clean exit.";
-              RAY_CHECK_OK(
-                  KillActor(actor_id, /*force_kill=*/false, /*no_restart=*/false));
-            }
-          }
-
-          absl::MutexLock lock(&actor_handles_mutex_);
-          // TODO(swang): Erase the actor handle once all refs to the actor
-          // have gone out of scope. We cannot erase it here in case the
-          // language frontend receives another ref to the same actor. In this
-          // case, we must remember the last task counter that we sent to the
-          // actor.
-          // TODO(ekl) we can't unsubscribe to actor notifications here due to
-          // https://github.com/ray-project/ray/pull/6885
-          auto callback = actor_out_of_scope_callbacks_.extract(actor_id);
-          if (callback) {
-            callback.mapped()(actor_id);
-          }
-        }));
-  }
-
-  return inserted;
+  return actor_manager_->SerializeActorHandle(actor_id, output, actor_handle_id);
 }
 
 Status CoreWorker::GetActorHandle(const ActorID &actor_id,
```

Comment:
This is actually used for the language frontend. I believe `CoreWorker` should be the only class that communicates with the `raylet.pyx` for simplicity, so I will just keep it here. Let me know if you want to expose `actorManager` to the language frontend.

Reply:
Hmm looking at the usage in `raylet.pyx`, it actually looks like it only needs access to the immutable info in the rpc::ActorHandle protobuf. Maybe we could just return that, so that we don't need to expose the C++ ActorHandle to raylet.pyx at all?

Reply:
Yeah it seems like language frontend only uses `inner` parts of actor handles. Will do this then!